rust/core-lib/src/editor.rs

```diff
@@ -691,171 +425,156 @@ impl Editor {
-        for region in view.sel_regions() {
-            let mut cursor = WordCursor::new(&self.text, region.end);
-            let (mut start, end) = cursor.select_word();
-
-            // if the word begins with '-', then it is a negative number
-            if start > 0 && self.text.byte_at(start - 1) == (b'-') {
-                start -= 1;
+        match cmd {
+            Delete { movement, kill } => {
+                let (delta, rope) = edit_ops::delete_by_movement(
+                    &self.text,
+                    regions,
+                    view.get_lines(),
+                    movement,
+                    view.scroll_height(),
+                    kill,
+                );
+                if let Some(rope) = rope {
+                    *kill_ring = rope;
+                }
+                if !delta.is_identity() {
+                    self.this_edit_type = EditType::Delete;
+                    self.add_delta(delta);
+                }
             }
-
-            let word = self.text.slice_to_cow(start..end);
-            if let Some(number) = word.parse::<i128>().ok().and_then(&transform_function) {
-                let interval = Interval::new(start, end);
-                builder.replace(interval, Rope::from(number.to_string()));
+            Backspace => {
+                let delta = edit_ops::delete_backward(&self.text, regions, config);
+                if !delta.is_identity() {
+                    self.this_edit_type = EditType::Delete;
+                    self.add_delta(delta);
+                }
             }
-        }
-
-        if !builder.is_empty() {
-            self.this_edit_type = EditType::Other;
-            self.add_delta(builder.build());
-        }
-    }
-
-    // capitalization behaviour is similar to behaviour in XCode
-    fn capitalize_text(&mut self, view: &mut View) {
-        let mut builder = DeltaBuilder::new(self.text.len());
-        let mut final_selection = Selection::new();
-
-        for &region in view.sel_regions() {
-            final_selection.add_region(SelRegion::new(region.max(), region.max()));
-            let mut word_cursor = WordCursor::new(&self.text, region.min());
-
-            loop {
-                // capitalize each word in the current selection
-                let (start, end) = word_cursor.select_word();
-
-                if start < end {
-                    let interval = Interval::new(start, end);
-                    let word = self.text.slice_to_cow(start..end);
-
-                    // first letter is uppercase, remaining letters are lowercase
-                    let (first_char, rest) = word.split_at(1);
-                    let capitalized_text =
-                        [first_char.to_uppercase(), rest.to_lowercase()].concat();
-                    builder.replace(interval, Rope::from(capitalized_text));
+            Transpose => {
+                let delta = edit_ops::transpose(&self.text, regions);
+                if !delta.is_identity() {
```

Comment:
we need to set `EditType::Transpose` here. It is probably worth doing another close read and make sure the edit type is correct in other places as well?

Reply:
Fixed. There was indeed another case, but I belive that got them all.

(Are we testing undo and redo behaviour?)

Reply:
There are a few tests around recording that exercise undo/redo, but overall not really. 😒